File: OneQuestionPerDay/350/intersection_of_two_arrays_ii.py
from typing import List, Dict
from collections import defaultdict


class Solution:
    def intersect(self, nums1: List[int], nums2: List[int]) -> List[int]:
        counter: Dict[int, int] = defaultdict(int)
        for n in nums1:
            counter[n] += 1

        res = []
        for n in nums2:
            if counter[n] > 0:
                res.append(n)
                counter[n] -= 1
        return res

File: OneQuestionPerDay/350/intersection_of_two_arrays_ii.go
package main

func intersect(nums1 []int, nums2 []int) []int {
	counter1 := make(map[int]int, 0)
	counter2 := make(map[int]int, 0)

	for _, n := range nums1 {
		counter1[n]++
	}
	for _, n := range nums2 {
		counter2[n]++
	}
	res := make([]int, 0)
	for k, v1 := range counter1 {
		if v2, ok := counter2[k]; ok {
			for i := 0; i < min(v1, v2); i++ {
				res = append(res, k)
			}
		}
	}
	return res
}

func min(x, y int) int {
	if x < y {
		return x
	}
	return y
}
